### ISSUE TITLE
Why do the results of the same image differ?

### ISSUE THREAD

**Issue by mulinhu** [facebookresearch/segment-anything#185]
Hi, I try to find all the objects in the image automatically. I used below code.
``` python
import numpy as np
import torch
import matplotlib.pyplot as plt
import cv2
import glob
def show_anns(anns,save_path):
    if len(anns) == 0:
        print(save_path)
        return
    sorted_anns = sorted(anns, key=(lambda x: x['area']), reverse=True)
    ax = plt.gca()
    ax.set_autoscale_on(False)
    polygons = []
    color = []
    for ann in sorted_anns:
        m = ann['segmentation']
        img = np.ones((m.shape[0], m.shape[1], 3))
        color_mask = np.random.random((1, 3)).tolist()[0]
        for i in range(3):
            img[:,:,i] = color_mask[i]
        ax.imshow(np.dstack((img, m*0.35)))
    plt.savefig(save_path)
        

import sys
sys.path.append("..")
from segment_anything import sam_model_registry, SamAutomaticMaskGenerator, SamPredictor

sam_checkpoint = "../sam_vit_h_4b8939.pth"
model_type = "vit_h"

device = "cuda"

sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
sam.to(device=device)

mask_generator = SamAutomaticMaskGenerator(sam)

files = glob.glob(fr"./*.jpg")
idx = 0
for file in files:
    image = cv2.imread(file)
   
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    print(f"image.shape:{image.shape}")
    plt.clf()
    plt.subplot(1,2,1)
    plt.imshow(image)
    
    plt.subplot(1,2,2)
    plt.imshow(image)
    masks = mask_generator.generate(image)
    print(fr"masks:{len(masks)}")
    show_anns(masks,fr"{idx}.png")
    idx += 1
    
```
However, I got this result:
![1](https://user-images.githubusercontent.com/118299117/231519894-54d971e7-f6ee-4546-9867-b08b14b05976.png)

But the demo's effect  is very good, you can see the below result
![1681316389260](https://user-images.githubusercontent.com/118299117/231520545-cab861b0-b505-461f-b2cd-f72b5d0024cd.png)

My biggest question is why the online effect is so good！！！
Have you used any other methods？

**Comment by AnasCHARROUD** [comment]
I remarked that also !!!!!!

**Comment by JordanMakesMaps** [comment]
There's also a number of parameters when initializing the model; you're currently using all of the default values.

**Comment by LedKashmir** [comment]
Pay attention to the show_anns() function, the line 'color_mask = np.random.random((1, 3)).tolist()[0]' may result in the difference in the same input, but I also don't know how to handle this problem.

**Comment by mulinhu** [comment]
The situation you mentioned may arise. But I think the probability of the same numerical value is very low. @LedKashmir

**Comment by Glisten-5481** [comment]
Hope to know what parameters are used in the online demo @HannaMao.

**Comment by Jordan-Pierce** [comment]
@huxycn have you seen an improvements doing any sort of image pre-processing? Obviously speed if resizing the image, but I've tried sharpening the image and that seems to help a little.

**Comment by Jack-bo1220** [comment]
same question! Is there any solution

**Comment by xmuzyyy** [comment]
Same question! The results can vary wildly. 
![5da1d642760dc16d489dacf86cc9276](https://user-images.githubusercontent.com/90310256/232327829-d4aabb99-e6e1-4316-ac67-cbb0a7fd65c0.png)
![976e7494716baf5bfe97a1ec9829b57](https://user-images.githubusercontent.com/90310256/232327911-045d2a43-4f6c-4963-9828-05ae82f80a11.png)

**Comment by nudlesoup** [comment]
following

**Comment by maheshs11** [comment]
@huxycn box_nms_thresh and crop_nms_thresh, how to set these so that i get only one mask and one bbox (no duplicate), so if set 0.5 it removes any overlap of more than 50 percent or less than 50?

**Comment by yong2khoo-lm** [comment]
+1 on this post.
i am getting different results too. The one on the web has way much better result.
It would be great to know the existing parameters. (or any additional processing)

> Additionally, generation can be automatically run on crops of the image to get improved performance on smaller objects, and post-processing can remove stray pixels and holes.

Wondering what is being done there...

**Comment by Akhp888** [comment]
+1 following

**Comment by dongjielie** [comment]
I tested 200 images, out of which 15 had poor segmentation results, but the online demo test results were excellent!!!, I really want to know why

**Comment by Young95K** [comment]
Same here

**Comment by helen1c** [comment]
+1 following

**Comment by HettyPatel** [comment]
+1 following

**Comment by chava100** [comment]
+1 following

**Comment by chava100** [comment]
I believe the inference done in the demo is on the onnx model.
when I ran examples an the onnx model results improved significantly I dont know why is it so but maybe this can help you.

**Comment by chava100** [comment]
> Hi @chava100 would you mind posting some screenshots, I know a lot of people would be interested.

The example I have is with a b_box prompt because I couldn't figure out how to do 'segment anything' when the model is exported to onnx.
A capture that shows results from the SAM pytorch model:
<img width="296" alt="image" src="https://user-images.githubusercontent.com/95570572/236332647-40b728f0-02e3-49ac-ac22-d3fd4bbf38b4.png">

A capture that shows results from SAM exported to onnx_quantized model:
<img width="295" alt="image" src="https://user-images.githubusercontent.com/95570572/236332949-a6ceab67-1d87-4dad-923f-81bb0729b035.png">

This is a capture from the demo and I cannot guarantee that the values of the b_box are exactly the same as in the other two images because I do not see how manual values can be entered but I tried to do it as close as possible:
<img width="458" alt="image" src="https://user-images.githubusercontent.com/95570572/236333375-f55733b9-ac1d-4bd7-93b1-34c2539dc26b.png">

**Comment by sliftist** [comment]
Also of note, quantizing the vit-h model gives pretty much the same mask result:

![image](https://github.com/facebookresearch/segment-anything/assets/52947011/2f543633-9616-452c-901e-fea34ca45af5)

**Comment by YutingZhang** [comment]
@sliftist I also often notice similar artifacts at the edges produced by the vit-h model. When using non-natural images (like 3D anime snapshots), these artifacts can sometimes become quite messy, even extending far from the edge with the vit-h model. However, the demo results look much cleaner in comparison.

Thank you also for the in-depth analysis using different feature maps with the same decoder model. It convincingly shows that either the model used for the demo (which appears to be improved) has not been released, or the input image was preprocessed somehow.

**Comment by idonahum1** [comment]
following +1

**Comment by BigBigMonsterProgramer** [comment]
Why do I get this result using onnx_quantized model? @chava100 
![output](https://github.com/facebookresearch/segment-anything/assets/105329568/4e09708e-8464-466e-986f-9d569cf6e05d)

**Comment by liren2515** [comment]
following +1

**Comment by jiangwei221** [comment]
> Why do I get this result using onnx_quantized model? @chava100 ![output](https://user-images.githubusercontent.com/105329568/260276741-4e09708e-8464-466e-986f-9d569cf6e05d.png) ![output2](https://user-images.githubusercontent.com/105329568/260276815-0017291f-8ae6-45a6-a915-fd4b311a7a9d.png)

Hi there!
I've met the same problem, do you know what causes the shifting/offset?
<img width="1321" alt="Screenshot 2023-09-28 at 10 25 40" src="https://github.com/facebookresearch/segment-anything/assets/25207535/10df2e33-f62b-47f2-9e06-f426a99784fa">

**Comment by phucbienvan** [comment]
following

**Comment by phucbienvan** [comment]
> > Why do I get this result using onnx_quantized model? @chava100 ![output](https://user-images.githubusercontent.com/105329568/260276741-4e09708e-8464-466e-986f-9d569cf6e05d.png) ![output2](https://user-images.githubusercontent.com/105329568/260276815-0017291f-8ae6-45a6-a915-fd4b311a7a9d.png)
> 
> Hi there! I've met the same problem, do you know what causes the shifting/offset? <img alt="Screenshot 2023-09-28 at 10 25 40" width="1321" src="https://user-images.githubusercontent.com/25207535/271367308-10df2e33-f62b-47f2-9e06-f426a99784fa.png"> Update: It's because I used a onnx model that is traced on 3:2 resolution, and applied it for 16:9 images.

hi @jiangwei221 
How to update a 3:2 model to 16:9?
thank you

**Comment by dddraxxx** [comment]
Does anyone come up with a better choice of parameter than default? how can we get the same result as demo

**Comment by Solskjaer-YES** [comment]
following

**Comment by WangSql** [comment]
follow

**Comment by yinwu33** [comment]
follow

**Comment by Dargonxzy** [comment]
following
